man/man3/history.3
.TH HISTORY 3NEMO "7 May 1992"
.SH NAME
get_history, put_history, app_history , ask_history, reset_history, ask_headline, set_headline \- data history utilities 
.SH SYNOPSIS
.nf
.B get_history(instr)
.B put_history(outstr)
.B app_history (s)
.B string *ask_history()
.B reset_history()
.PP
.B void set_headline(s)
.B string ask_headline()
.PP
.B stream instr, outstr;
.B string s;
.fi
.SH DESCRIPTION
\fIhistory\fP is a general mechanism to keep track of data history,
and also store random user-supplied comments to data in
\fIfilestruct(5NEMO)\fP format. One history database at a time
can be managed, history from a file can be added to this, and
it can be written out to a file. Whenever a NEMO command is run,
the history is initialized with the actual command line supplied. 
History from datafiles can be appended to this.
.PP
\fIget_history()\fP appends history from a file
\fBinstr\fP (assumed to be in \fIfilestruct(5NEMO)\fP format) to 
the internal history buffer. This is usually
one of the first things a program should do before other input is
attempted, as the data history is normally prepended to the data. 
.PP
A program can add (append) arbitrary history by calling
\fIapp_history()\fP, where \fBs\fP points to the string to be added.
.PP
\fIput_history()\fP writes the current history to an output file 
\fBoutstr\fP.
If any headline had been set, it will also be written.
.PP
\fIask_history()\fP returns a pointer to a NULL terminated array of
strings that contain the full history status.
.PP
\fIset_headline()\fP is an old way to add user supplied to the history
database. Only one string can be hold here. It is added to the data history
when \fIput_history\fP is called.
.PP
\fIask_headline()\fP returns the headline string, which would be output
.SH FILESTRUCT FORMAT
History items are character strings with tagname \fBHistory\fP, 
headline items are character strings with tagname \fBHeadline\fP.
They are both treated equal: if a headline is present, it is also
written via \fIput_history\fP.
.SH FILES
.nf
.ta +2.0i
~/src/kernel/io   	history.[ch]
~/inc                	history.h
.fi
.SH SEE ALSO
getparam(3NEMO), tsf(1NEMO), rsf(3NEMO)
.SH AUTHOR
Peter Teuben
.SH UPDATE HISTORY
.nf
.ta +1.0i +4.0i
9-mar-88	V1.0 created		Peter Teuben
1-jun-88	V1.1 adapted to new filestruct, renamed names	PJT
14-jun-88	V1.3 integrate at IAS, changed _hist to _history	JEB
9-oct-90	V1.8b upgraded manual page	PJT
7-may-92	V1.9 documented name change add_ to app_	PJT
.fi


usr/jcl/glnemo2/plugins/nemolight/history.c
/*
 * HISTORY.C: handle history and headline items on the toplevel 
 *            of a structured binary file.
 *
 *	 9-Mar-88	V1.0	Created					PJT
 *	 1-Jun-88	V1.1	renamed read/write to get/put		PJT
 *	 7-jun-88	V1.2	installed in libT with getparam();
 *                              dynamic HISTORY				PJT
 *      14-Jun-88       V1.3    integrated with IAS NEMO		JEB
 *	15-Dec-88	V1.4	allow creation from scratch		PJT
 *	15-jul-89	V1.5	added reset_history(), fixed order	PJT
 *				of Headline's and History's
 *	29-sep-89	V1.6	added item= keyword                     PJT
 *	 8-dec-89	V1.7	moved some stuff to history.h, no error PJT
 *	10-sep-90       V1.8    << moved history_level around >>	PJT
 *	 9-oct-90	V1.8a   documentation upgraded			PJT
 *      17-jul-91       V1.9    extra option skip=                      PJT
 *
 *	 9-apr-92	V2.0    add_history -> app_history		PJT
 *				because of conflicting libreadline.a
 *	20-feb-94	V2.1    ansi
 *	13-mar-95       V2.2	app_history() now makes new copy of history 
 *				string -- hmm, this bug didn't show up until
 *				linux! - also did the prototypesm
 *
 *	 6-may-95       V2.3    (hisf) allow item=headline         	PJT
 *	21-may-01	V2.4    history <old history_level> back here   PJT
 *       1-aug-02       V2.5    increased memory a bit 256->1024        
 *                              and free up memory in reset_history     PJT
 *      28-dec-04       V2.5a   tag headline properly when free'd       PJT
 *
 *  ToDo: not all local variables free up memory, despite that some
 *        have clearly come from allocate'd memory. 
 */

#include <stdinc.h>
#include <strlib.h>
#include <filestruct.h>
#include <history.h>

/* 
 * The datastructure to store history is a simple array of pointers to
 * history strings (in safe memory we may hope).
 * The array is declared static, but you may also think of
 * a dynamically allocated linked list in the future.
 *				-- sure, and just how will that work?  JEB
 */

#ifndef MAXHIST
# define MAXHIST 1024			/* max size of history array        */
#endif

local int nhist = 0;			/* count history data stored so far */
local string histbuf[MAXHIST+1];	/* history string array             */
local string headline = NULL;		/* last headline read in            */
local bool freeup[MAXHIST+1];           /* if space should be free'd        */

int history = 1;               /* 1=history is auto-mode  0=no history done */
                               /* this item should eventually disappear     */
                               /* it is defined by the user interface,      */
                               /* see getparam.c                            */

/*
 * GET_HISTORY: read history and headline info into local buffer;
 * returns current number of history items
 */

int get_history(stream instr)
{
    for(;;) {                   		/* loop reading input data  */
	if (get_tag_ok(instr, HeadlineTag)) {
	    headline = get_string(instr, HeadlineTag);
	    dprintf(5, "get_history: headline = %s\n", headline);
	} else if (get_tag_ok(instr, HistoryTag)) {
	    if (nhist > MAXHIST) {
		warning("get_history: no more history saved; MAXHIST=%d",
                        MAXHIST);
		return MAXHIST;
	    }
	    histbuf[nhist] = get_string(instr, HistoryTag);
	    dprintf(5, "get_history: histbuf[%d] = %s\n",
		    nhist, histbuf[nhist]);
	    freeup[nhist] = FALSE;
	    nhist++;
	} else
            break;                          /* done with reading loop */
    }
    return nhist;
}

/*
 * PUT_HISTORY:  write current history and headline data to output.
 *               always returns 0 ...
 */

int put_history(stream outstr)
{				
    int i;

    if (!history) {
    	dprintf(5, "put_history: history data suppressed\n");    
        return 0;
    }
    if (headline != NULL) {
	dprintf(5, "put_history: headline = %s\n", headline);
	put_string(outstr, HeadlineTag, headline);
    }
    dprintf(5, "put_history: writing %d history items\n", nhist);
    for (i = 0; i < nhist; i++) {
	dprintf(5, "             histbuf[%d] = %s\n", i, histbuf[i]);
    	put_string(outstr, HistoryTag, histbuf[i]);
    }
/*  nhist = 0;			// reset counter -- why? JEB */
    return 0;
}

/* 
 * APP_HISTORY: add to history
 *              returns current length of history list
 *
 */

int app_history(string s)
{
    permanent bool warned = FALSE;

    if (nhist > MAXHIST) {
	if (!warned) warning("app_history: too much history");
        warned = TRUE;
	return nhist;
    } else if (s == NULL || streq(s, "")) {
	dprintf(1, "app_history: null history string\n");
	return nhist;
    }
    histbuf[nhist] = scopy(s);
    freeup[nhist] = TRUE;
    dprintf(9,"app_history: histbuf[%d] = %s\n", nhist, s);
    nhist++;
    return nhist;
}

/*
 * RESET_HISTORY: reset the history list, and also free up all memory
 */

void reset_history()
{
  int i;
  for (i=0; i<nhist; i++)
    if (freeup[i]) free(histbuf[i]);
  if (headline) {
    free(headline);
    headline=0;
  }
  nhist = 0;
}

/*
 * SET_HEADLINE: set new value for headline string.
 */

void set_headline(string s)
{
    headline = scopy(s);
}

/*
 * ASK_HEADLINE: enquire about value of headline.
 */

string ask_headline()
{
    return headline;
}

/*
 * ASK_HISTORY: enquire about history data.
 */

string *ask_history()
{
    if (nhist > MAXHIST)
	error("ask_history: too much history");
    histbuf[nhist] = NULL;			/* NULL-term the hist array */
    return histbuf;
}

#ifdef TOOLBOX
			/* PROGRAM: hisf */
#include <getparam.h>

string defv[] = {
    "in=\n		Input file name",
    "out=\n		If given, output history",
    "item=all\n		Show which items",
    "history=\n		If given, add this text to history",
    "skip=1\n           Number to skip before display (0 or 1)",
    "VERSION=2.3\n	6-may-95 PJT",
    NULL,
};

string usage = "list history items from a binary structured file";

string iname,oname;			/* input/output file  name */
stream instr,outstr;

void nemo_main()
{
    int i, skip;
    int item = -1;
    bool Qall = FALSE, Qhead=FALSE;

    iname = getparam("in");
    oname = getparam("out");
    if (streq(getparam("item"),"all")) 
        Qall = TRUE;
    else if (streq(getparam("item"),"headline"))
        Qhead = TRUE;
    else
        item = getiparam("item");       /* get a particular history item */
    skip = getiparam("skip");
    if (skip<0) error("skip=%d cannot be negative",skip);

    if (hasvalue("in")) {
        instr = stropen(iname, "r");                /* open input file */
        get_history(instr);                         /* get old history */
        if (hasvalue("history"))
	    app_history(getparam("history"));       /* add new */
        for (i = skip; i < nhist; i++) 
            if (Qall || i==item) {
                dprintf(1,"%d: ",i);
                printf("%s\n", histbuf[i]);     /* list old history */
            }
        if (Qhead && headline) printf("%s\n",headline);     /* list headline */
        strclose(instr);
    } else if (hasvalue("history")) {
        nhist--;				/* delete first one */
        app_history(getparam("history"));           /* add new from scratch */
    }

    if (hasvalue("out")) {                       /* see if output needed */
        outstr = stropen(oname, "w");               /*  open file */
        put_history(outstr);                        /*  and dump history */
        strclose(outstr);                           
    }
}

#endif
